pyautocausal/orchestration/pipeline_graph.py

```
@@ -13,39 +14,124 @@ class PipelineGraph(ExecutableGraph):
+        self.input_class.check_presence_of_required_fields()
+        
+        #TODO: Remove any BaseNode
+
+        # Add nodes from input data
         self.add_initial_nodes_from_dict(input_data.to_dict())
-    
+        
+        # Execute the graph
+        self.execute_graph()
+        
+        return self
+        
+    def add_node(self, node: Node):
+        """Add a node to the graph and track it in the node index.
+        
+        Args:
+            node: Node to add to the graph
+            
+        Raises:
+            ValueError: If a node with the same name already exists
+        """
+        if node.name in self.node_index:
+            raise ValueError(f"Node with name '{node.name}' already exists in the graph")
+        
+        super().add_node(node)
+        self.node_index[node.name] = node
+
     def add_initial_nodes_from_dict(self, initial_node_dict: dict):
         """Create initial nodes from input dictionary"""
+        
         for node_name, node_output in initial_node_dict.items():
             node = Node(
-                name=node_name, 
+                name=node_name,
                 graph=self, 
                 action_function=lambda x=node_output: x
             )
+            
+            self.connect_initial_nodes(node)
+
+    
+    def connect_initial_nodes(self,initial_node: Node):
+        """ Connect a node to the starter nodes """
+        for node_name, starter_node in self.starter_nodes.items():
+            #TODO: Make this more flexible using a result class that gets passed through the graph
+            starter_node.add_predecessor(initial_node, argument_name = "df")
+
+    def add_branch(self, steps: list[tuple[str, Node]], predecessor: str = None):
+        """Add a branch of nodes to the graph.
+        
+        Args:
+            steps: List of (name, node) tuples defining the branch
+            predecessor: Name of node to connect branch to (optional)
+        """
+        if not steps:
```

probably should throw an error unless we'd ever want to call this with an empty list